File: src/main/java/de/wittenbude/expandify/models/db/Playlist.java
package de.wittenbude.expandify.models.db;

import de.wittenbude.expandify.models.pojos.Followers;
import de.wittenbude.expandify.models.pojos.Image;
import de.wittenbude.expandify.models.pojos.PlaylistTrack;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.DocumentReference;
import se.michaelthelin.spotify.enums.ModelObjectType;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

@Data
@NoArgsConstructor
@Document(collection = "playlists")
public class Playlist {

    @Id
    private String id;
    private Boolean collaborative;
    private String description;
    private Map<String, String> externalUrls;
    private Followers followers;
    private String href;
    private String spotifyId;
    private Image[] images;
    private String name;
    @DocumentReference(lazy = true)
    private SpotifyUserPublic owner;
    private Boolean publicAccess;
    private String snapshotId;
    private List<PlaylistTrack> tracks;
    private Integer totalTracks;
    private ModelObjectType type;
    private String uri;

    public Playlist(se.michaelthelin.spotify.model_objects.specification.Playlist playlist, List<PlaylistTrack> playlistTracks) {
        this.collaborative = playlist.getIsCollaborative();
        this.externalUrls = playlist.getExternalUrls().getExternalUrls();
        this.href = playlist.getHref();
        this.spotifyId = playlist.getId();
        this.images = Arrays.stream(playlist.getImages()).map(Image::new).toArray(Image[]::new);
        this.name = playlist.getName();
        this.owner = new SpotifyUserPublic(playlist.getOwner());
        this.publicAccess = playlist.getIsPublicAccess();
        this.snapshotId = playlist.getSnapshotId();
        this.tracks = playlistTracks;
        this.uri = playlist.getUri();
        this.description = playlist.getDescription();
        this.followers = new Followers(playlist.getFollowers());
        this.totalTracks = playlist.getTracks().getTotal();

        this.id = this.spotifyId + this.snapshotId;
    }
}

File: src/main/kotlin/de/wittenbude/exportify/models/Playlist.kt
@file:Suppress("PLUGIN_IS_NOT_ENABLED")

package de.wittenbude.exportify.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import se.michaelthelin.spotify.enums.ModelObjectType
import se.michaelthelin.spotify.model_objects.specification.PlaylistSimplified
import se.michaelthelin.spotify.model_objects.specification.Playlist as ApiPlaylist

@Serializable
data class Playlist(
    val collaborative: Boolean? = null,
    val description: String? = null,
    val externalUrls: Map<String, String>? = null,
    val followerHref: String? = null,
    val followerTotal: Int? = null,
    val href: String? = null,
    @SerialName("_id") val id: String,
    val images: List<String>? = null,
    val name: String,
    val owner: String? = null,
    val publicAccess: Boolean? = null,
    val snapshotId: String? = null,
    val tracks: List<String>? = null,
    val type: ModelObjectType? = null,
    val uri: String? = null,
)


fun PlaylistSimplified.convert(tracks: List<Track>?): Playlist {
    return Playlist(
        collaborative = isCollaborative,
        externalUrls = externalUrls?.externalUrls,
        href = href,
        id = id,
        images = images?.map { it.url },
        name = name,
        owner = owner?.id,
        publicAccess = isPublicAccess,
        snapshotId = snapshotId,
        tracks = tracks?.map { it.id },
        type = type,
        uri = uri
    )
}

fun ApiPlaylist.convert(tracks: List<Track>?): Playlist {
    return Playlist(
        this.isCollaborative,
        description,
        externalUrls?.externalUrls,
        followers?.href,
        followers?.total,
        href,
        id,
        images?.map { it.url },
        name,
        owner?.id,
        isPublicAccess,
        snapshotId,
        tracks?.map { it.id },
        type,
        uri
    )
}